### PR TITLE
[GSoC] Simulation Object Restructure: First Objective

### PR DESCRIPTION
**Pull request by AyushiDaksh**
### :pencil: Description

**Type:** 🌞 `GSoC`

This is my solution to the first objective of the GSoC'23 project: Simulation Object Restructure.

Here, I have created a separate class called ConvergenceStrategy which is called internally by the Simulation class.

### :pushpin: Resources
[GSoC'23 Page](https://summerofcode.withgoogle.com/)
[GSoC'23 Ideas Page](https://tardis-sn.github.io/gsoc_2023/ideas/)

### :vertical_traffic_light: Testing

How did you test these changes?

- [x] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)

### :ballot_box_with_check: Checklist

- [x] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

### CODE REVIEW

**Review by atharva-2001 — COMMENTED** [review]
Thanks for creating this pull request! Did tests pass with your changes(assuming you provided a path to the reference data)?

File: tardis/simulation/base.py [hunk 1]
```diff
@@ -216,74 +200,10 @@ def estimate_t_inner(
...
-
     def advance_state(self):
         """
         Advances the state of the model and the plasma for the next
-        iteration of the simulation. Returns True if the convergence criteria
-        are met, else False.
-
-        Returns
-        -------
-            converged : bool
```
Comment (sonachitchyan) [inline tardis/simulation/base.py:286]: you should still have what the function returns(or what it updates in this case)

**Review by sonachitchyan — CHANGES_REQUESTED** [review]
Thank you for the PR! There is one small thing that I commented about one of the docstrings. Everything else looks great. As a next step, can you add a simple unit test and be sure that the documentation shows explicitly shows how the convergence plot is treated?

File: tardis/simulation/base.py (hunk 1 above)
Comment (AyushiDaksh) [reply tardis/simulation/base.py:286]: Fixed this!

File: tardis/simulation/base.py [hunk 2]
```diff
@@ -441,11 +340,11 @@ def run(self):
...
             )
             self.iterate(self.no_of_packets)
-            self.converged = self.advance_state()
+            self.advance_state()
             if hasattr(self, "convergence_plots"):
                 self.convergence_plots.update()
             self._call_back()
-            if self.converged:
+            if self.convergence_strategy.converged:
                 if self.convergence_strategy.stop_if_converged:
```
Comment (Rodot-) [inline tardis/simulation/base.py:348]: Perhaps this should be a single conditional and the convergence_strategy should have a separate single attribute that tells it to stop or not based on whether or not stop_if_converged is true.

Comment (AyushiDaksh) [reply tardis/simulation/base.py:348]: I see. I'll add a new @property "should_stop = self.converged and self.stop_if_converged" in the ConvergenceStrategy class. We'll only check this property in the simulation class above.

If you can think of a better name than "should_stop", please let me know

Comment (AyushiDaksh) [reply tardis/simulation/base.py]: @Rodot- I've handled this, please review

I also wanted to ask, what test cases can I add for this PR?

**Pull request closed without merging**